worker/include/RTC/RtpStreamSend.hpp

```diff
@@ -45,14 +48,19 @@ namespace RTC
 		public:
 			~StorageItemBuffer();
 
-			StorageItem* Get(uint16_t seq);
+			StorageItem* Get(uint16_t seq) const;
 			bool Insert(uint16_t seq, StorageItem* storageItem);
 			bool Remove(uint16_t seq);
 			void Clear();
+			StorageItem* GetOldest() const;
 
 		private:
-			uint16_t startSeq{ 0 };
-			std::deque<StorageItem*> buffer;
+			void InvalidateOldest(uint16_t prevOldest);
+			static size_t GetBucketIndex(uint16_t seq);
+			static size_t GetPositionInBucket(uint16_t seq);
+
+			int32_t oldestSeq{ -1 };
+			std::array<Utils::Bucket<StorageItem, BucketSize>*, BucketsCount> buckets{};
```

Comment:
I don't understand the idea behind 2 arrays/levels of indirection, can you explain why is this needed?

Reply:
@nazar-pc It's working like [virtual memory](https://en.wikipedia.org/wiki/Virtual_memory). I create [visualization](https://gist.github.com/Hartigan/61e194fcd561c14c0e2b8086327034f9). ```Buckets/Pages``` are reusable in thread.

Reply:
I think I understand what, I don't understand why though. It worked fine before (minus apparent bug somewhere) and used much less memory than this approach after quick glance.

Reply:
```std::deque``` allocations depends on [platform](https://bugs.chromium.org/p/chromium/issues/detail?id=674287). For ```MSVC``` it's working like linked list with frequent allocation. My approach uses buffers with size ```8 * 256 = 2048 bytes``` and ```8 * 256 = 2048 bytes``` for bucket's ```std::array```.
```std::deque``` buffer sizes :
- MSVC: 16 bytes or element size if it's bigger(16 bytes for our case)
- GCC: 512 bytes or element size if it's bigger(512 bytes for our case)
- Clang: element_size < 256 ? 4096 : element_size * 16(4096 bytes for our case).

So, my approach uses comparable memory size with ```std::deque``` for GCC/CLang, but consumes much less allocations.

Also, ```std::deque``` it's linked list of buffers, so you need to dereference the pointer multiple times for ```Get``` by index(depends on buffers number and implementation). In my approach we need only 2 dereference operations.

Reply:
https://github.com/Shmoopty/veque looks promising, I can write `meson.build` for it.

Reply:
```Boost``` provides [circular_buffer_space_optimized](https://www.boost.org/doc/libs/1_78_0/doc/html/circular_buffer.html), but need to implement ```seq``` to ring buffer ```idx``` and import ```boost```.

Reply:
Anyway, current implementation on ```deque``` has not obvious translation from ```seq``` to in ```deque``` position and may contain hidden bugs.